File: c++/recursive/inorder.cpp
#include<bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x)  {}
};


void inorder(TreeNode *root){
    if(!root)return ;
    inorder(root->left);
    cout<<root->val<<" ";
    inorder(root->right);
}

int main()
{
    TreeNode * root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left = new TreeNode(4);
    root->left->right = new TreeNode(5);
    root->right->left = new TreeNode(6);
    root->right->right = new TreeNode(7);

    inorder(root);
}

File: trees/java/iterative/inorder.java
package trees.java.iterative;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

class NodeInfo{
    
    TreeNode node;
    boolean visited;

    public NodeInfo(TreeNode node,boolean visited){
        this.node = node;
        this.visited = visited;
    }

}


public class inorder {
    List<Integer> ans ;
    TreeNode root;

    public inorder(TreeNode root){
        ans = new ArrayList<>();
        this.root = root;
    }

    public void traverseInOrder(TreeNode root){
        Stack<NodeInfo> st = new Stack<>();
        st.push(new NodeInfo(root, false));

        while(!st.isEmpty()){
            NodeInfo temp = st.peek();
            if(temp.node==null){
                st.pop();
                continue;
            }
            if(temp.visited){
                st.pop();
                ans.add(temp.node.val);
                st.push(new NodeInfo(temp.node.right, false));
            }else{
                temp.visited=true;
                st.push(new NodeInfo(temp.node.left, false));
            }
        }

    }

    public void print(){
        System.out.print("In order traversal : ");
        for(int i:ans){
            System.out.print(i+" ");
        }
        System.out.println();
    }

    public void operation(){
        traverseInOrder(root);
        print();
    }

}
